src/test/java/io/axoniq/axonserver/connector/DoOnFirstResultStreamTest.java

```diff
@@ -0,0 +1,158 @@
+
+import static java.util.Arrays.asList;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+/**
+ * Tests {@link DoOnFirstResultStream}.
+ */
+class DoOnFirstResultStreamTest {
+
+    private static final String ELEMENT = "element";
+    private static final boolean CLOSED = true;
+
+    private ResultStream<String> delegate;
+    private Consumer<String> doOnFirstCallback;
+    private DoOnFirstResultStream<String> testSubject;
+
+    @BeforeEach
+    void setUp() throws InterruptedException {
+        //noinspection unchecked
+        delegate = mock(ResultStream.class);
+        when(delegate.peek()).thenReturn(ELEMENT);
+        when(delegate.nextIfAvailable()).thenReturn(ELEMENT);
+        when(delegate.nextIfAvailable(anyLong(), any(TimeUnit.class))).thenReturn(ELEMENT);
+        when(delegate.next()).thenReturn(ELEMENT);
+        when(delegate.isClosed()).thenReturn(CLOSED);
+        when(delegate.getError()).thenReturn(Optional.empty());
+        //noinspection unchecked
+        doOnFirstCallback = mock(Consumer.class);
+        testSubject = new DoOnFirstResultStream<>(delegate, doOnFirstCallback);
+    }
+
+    @Test
+    void testOnFirstCallbackInvokedWhenPeeked() {
+        assertEquals(ELEMENT, testSubject.peek());
+        verify(doOnFirstCallback).accept(ELEMENT);
+
+        assertEquals(ELEMENT, testSubject.peek());
+        verifyNoMoreInteractions(doOnFirstCallback);
+        verify(delegate, times(2)).peek();
+    }
+
+    @Test
+    void testOnFirstCallbackInvokedWhenNextIfAvailable() {
+        assertEquals(ELEMENT, testSubject.nextIfAvailable());
+        verify(doOnFirstCallback).accept(ELEMENT);
+
+        assertEquals(ELEMENT, testSubject.nextIfAvailable());
+        verifyNoMoreInteractions(doOnFirstCallback);
+        verify(delegate, times(2)).nextIfAvailable();
+    }
+
+    @Test
+    void testOnFirstCallbackInvokedWhenNextIfAvailableWithTimeout() throws InterruptedException {
+        assertEquals(ELEMENT, testSubject.nextIfAvailable(1, TimeUnit.SECONDS));
+        verify(doOnFirstCallback).accept(ELEMENT);
+
+        assertEquals(ELEMENT, testSubject.nextIfAvailable(1, TimeUnit.SECONDS));
+        verifyNoMoreInteractions(doOnFirstCallback);
+        verify(delegate, times(2)).nextIfAvailable(1, TimeUnit.SECONDS);
+    }
+
+    @Test
+    void testOnFirstCallbackInvokedWhenNext() throws InterruptedException {
+        assertEquals(ELEMENT, testSubject.next());
+        verify(doOnFirstCallback).accept(ELEMENT);
+
+        assertEquals(ELEMENT, testSubject.next());
+        verifyNoMoreInteractions(doOnFirstCallback);
+        verify(delegate, times(2)).next();
+    }
+
+    @Test
+    void testDelegationOfNonConsumableMethods() {
+        assertTrue(testSubject.isClosed());
+        assertEquals(Optional.empty(), testSubject.getError());
+        testSubject.close();
+        testSubject.onAvailable(null);
+
+        verify(delegate).isClosed();
```

Comment:
Just to be sure, you would want to verify that each invocation is delegated to the right method on the delegate. Maybe just put each of the `verify` right after the method that triggered it? Alternatively, verify the invocations `InOrder`